### PR TITLE
API Security

### LINKED ISSUE #272: API security
For non-sidecar/host-local use cases, we need to secure the REST API. Creating a ticket to track the work required here. Some immediate thoughts:

1) Enable TLS on server-side.
2) Investigate options for API auth/n and auth/z.

For auth/z, we could probably get away with:

1) Read-only, data access
2) Read-write, data access
3) Read-write, data+policy access

Alternatively, we could use Rego to define the API auth/z policy which would give complete flexibility.

### PR DESCRIPTION
**Pull request by tsandall**
These changes close #272 by adding support for authentication and authorization.

0ede160 contains the main additions to perform token-based authentication and authorization in OPA's server.

### CODE REVIEW

File: site/documentation/how-does-opa-work/index.md [hunk 1]
```diff
@@ -159,7 +159,7 @@ A policy file must contain a single package declaration, which defines the path
 
 In contrast to base documents, virtual documents embody the results of evaluating the rules included in policy modules. Virtual documents are computed when users publish new policy modules, update existing modules, run queries, and when any relevant base document is published or updated. Rules allow policy authors to write questions with yes-no answers (that is, predicates) and to generate structured values from raw data found in base documents as well as from intermediate data found in other virtual documents.
 
-### The data Document
+### <a name="the-data-document"></a> The data Document
```
Comment (mikol) [inline site/documentation/how-does-opa-work/index.md]: Should “data” be in backticks to emphasize its role? As in: ``​`data`​``

File: site/documentation/references/security/index.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+## API Security
+
+Securing the API involves configuring OPA to use TLS, authentication, and
+authorization so that:
+
+- Traffic between OPA and clients is encrypted.
+- Clients verify the OPA API endpoint identity.
+- OPA verifies client identities.
+- Clients are only granted access to specific APIs or sections of [The data Document](/documentation/how-does-opa-work#the-data-document).
```
Comment (mikol) [inline site/documentation/references/security/index.md]: ``​`data`​``

File: site/documentation/references/security/index.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+```http
+HTTP/1.1 401 Unauthorized
+Content-Type: application/json
+```
+
+```json
+{
+  "code": "unauthorized",
+  "message": "request rejected by administrative policy"
```
Comment (mikol) [inline site/documentation/references/security/index.md]: Should messages use sentence case? As in: `"Request rejected..."`

File: site/documentation/references/security/index.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+At a minimum, the authorization policy should grant access to a special root
+identity:
+
+```ruby
+package system.authz
+
+default allow = false          # Reject requests by default.
+
+allow {                        # Allow request if...
+    input.identity = "secret"  # Identity is the secret root key.
```
Comment (mikol) [inline site/documentation/references/security/index.md]: To mitigate potential confusion over “assignment”, perhaps: `"secret" = input.identity`?

File: site/documentation/references/security/index.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+allow {                        # Allow request if...
+    tokens[input.identity]     # Identity exists in "tokens".
+}
+
+# Tokens may defined in policy or pushed into OPA as data.
```
Comment (mikol) [inline site/documentation/references/security/index.md]: At first, I scrolled back up to see if `tokens` had been defined before. Maybe place this directly after the import statements?

File: site/documentation/references/security/index.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+tokens = {
+    "my-secret-token-foo": {
+        "roles": ["admin"],
+    },
+    "my-secret-token-bar": {
+        "roles": ["service-1"],
+    },
+    "my-secret-token-baz": {
+        "roles": ["service-2", "service-3"],
+    },
```
Comment (mikol) [inline site/documentation/references/security/index.md]: Trailing comma… Nice that it’s legal in Rego. Possible point of confusion if this code is copy-pasted to a JSON document, then pushed to OPA.

File: site/documentation/how-does-opa-work/index.md (hunk 1 above)
Comment (tsandall) [inline site/documentation/how-does-opa-work/index.md]: Yup, will fix.

File: site/documentation/references/security/index.md (hunk 2 above)
Comment (tsandall) [inline site/documentation/references/security/index.md]: Will fix.

File: site/documentation/references/security/index.md (hunk 4 above)
Comment (tsandall) [inline site/documentation/references/security/index.md]: Will fix.

File: site/documentation/references/security/index.md (hunk 3 above)
Comment (tsandall) [inline site/documentation/references/security/index.md]: None of the other messages currently do. Will create a ticket to track this.

File: site/documentation/references/security/index.md (hunk 6 above)
Comment (tsandall) [inline site/documentation/references/security/index.md]: Will fix.

File: site/documentation/references/security/index.md (hunk 5 above)
Comment (tsandall) [inline site/documentation/references/security/index.md]: Will fix. This is just habit as I tend to write the policies top-down.

**Pull request merged**